transport/src/main/java/io/netty/channel/CoalescingBufferQueue.java

```diff
@@ -0,0 +1,191 @@
+            return Unpooled.EMPTY_BUFFER;
+        }
+
+        ByteBuf toReturn = null;
+        CompositeByteBuf composite = null;
+        int originalBytes = bytes;
+        for (;;) {
+            Object entry = promisesAndBufs.poll();
+            if (entry == null) {
+                break;
+            }
+            if (entry instanceof ChannelPromise) {
+                aggregatePromise.addListener(new ChannelPromiseNotifier((ChannelPromise) entry));
+                continue;
+            }
+            ByteBuf entryBuffer = (ByteBuf) entry;
+            int entryReadableBytes = entryBuffer.readableBytes();
+            if (bytes > 0 || entryReadableBytes == 0) {
+                if (toReturn == null) {
+                    if (entryReadableBytes > bytes) {
+                        promisesAndBufs.addFirst(entryBuffer.retain());
+                        toReturn = entryBuffer.readSlice(bytes);
+                        bytes = 0;
+                    } else if (entryReadableBytes == bytes) {
+                        toReturn = entryBuffer;
+                        bytes = 0;
+                    } else {
+                        composite = alloc.compositeBuffer(promisesAndBufs.size() + 1);
+                        composite.addComponent(entryBuffer);
+                        composite.writerIndex(entryReadableBytes);
+                        bytes -= entryReadableBytes;
+                        toReturn = composite;
+                    }
+                } else if (entryReadableBytes > bytes) {
+                    ByteBuf slice = entryBuffer.readSlice(bytes);
+                    composite.addComponent(slice);
+                    composite.writerIndex(composite.writerIndex() + slice.readableBytes());
+                    // The remaining buffer is kept as a retained slice of the original.
+                    promisesAndBufs.addFirst(entryBuffer.slice().retain());
+                    bytes = 0;
+                } else {
+                    if (composite == null) {
+                        composite = alloc.compositeBuffer(promisesAndBufs.size() + 1);
+                        composite.addComponent(toReturn);
+                        composite.writerIndex(toReturn.readableBytes());
+                        toReturn = composite;
+                    }
+                    composite.addComponent(entryBuffer);
+                    composite.writerIndex(composite.writerIndex() + entryReadableBytes);
+                    bytes -= entryReadableBytes;
+                }
+            } else {
+                // Dequeued a non-empty entry we did not consume so add it back at the head of the queue.
+                promisesAndBufs.addFirst(entry);
+                break;
+            }
+        }
+        readableBytes -= originalBytes - bytes;
+        assert readableBytes >= 0;
+        return toReturn;
+    }
+
+    /**
+     * The number of readable bytes.
+     */
+    public int readableBytes() {
```

Comment:
Maybe change this to `isEmpty()` or add an `isEmpty()` because we support empty buffers?